### PR TITLE
Fix part of #42: Exploration player base (Part 1) #42

### PR DESCRIPTION
**Pull request by rt4914**
## Explanation

This PR contains basic activities/fragment structure for exploration player. Currently the controllers are not functional which will be covered in subsequent PRs. Also this PR also contains one class for `FakeDataProvider`, this class is for couple of PRs only and later on it can be removed once `domain` / `model` layers are finished

Project structure:
Screenshot 2019-09-13 at 1 49 59 PM

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is **assigned** to an appropriate reviewer.

### CODE REVIEW

**Comment by rt4914** [comment]
@veena14cs @nikitamarysolomanpvt @BenHenning PTAL.

Also, @veena14cs @nikitamarysolomanpvt use this branch as base branch for your subsequent work on exploration-player.

File: app/src/main/java/org/oppia/app/FakeDataProvider.kt [hunk 1]
```diff
@@ -0,0 +1,27 @@
+package org.oppia.app
+
+import com.squareup.moshi.JsonAdapter
+import com.squareup.moshi.Moshi
+import org.oppia.data.backends.gae.model.GaeExplorationContainer
+
+// Remove this file at later stage once domain and model module are working.
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/FakeDataProvider.kt]: I actually think we should be faking data at a higher level than this (e.g. in domain logic controllers). We should spec out those interfaces and submit them with stubs that return structured fake data, then directly rely on those for the UI.

What fake data do we need for this PR? Can we just check in the structure and hook up to domain controllers in a subsequent PR?

File: app/src/main/java/org/oppia/app/player/exploration/ExplorationActivity.kt [hunk 2]
```diff
@@ -0,0 +1,16 @@
+package org.oppia.app.player.exploration
+
+import android.os.Bundle
+import org.oppia.app.activity.InjectableAppCompatActivity
+import javax.inject.Inject
+
+/** The starting point for exploration*/
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/player/exploration/ExplorationActivity.kt]: Nit: Documentation should include ending punctuation. And, there should be a space before the terminating ``*/``.

File: app/src/main/java/org/oppia/app/player/exploration/ExplorationActivity.kt [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+class ExplorationActivity : InjectableAppCompatActivity() {
+  @Inject lateinit var explorationActivityController: ExplorationActivityController
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/player/exploration/ExplorationActivity.kt]: I started calling these presenters in #106. Let's rename them to be presenters in this PR, too, to avoid the ambiguity of 'controller' here.

File: app/src/main/java/org/oppia/app/player/exploration/ExplorationActivityController.kt [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+import javax.inject.Inject
+
+/** The controller for [ExplorationActivity]. */
+@ActivityScope
+class ExplorationActivityController @Inject constructor(private val activity: AppCompatActivity) {
+  fun handleOnCreate() {
+    activity.setContentView(R.layout.exploration_activity)
+    activity.supportFragmentManager.beginTransaction().add(
+      R.id.exploration_fragment_placeholder,
+      ExplorationFragment()
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/player/exploration/ExplorationActivityController.kt]: I noticed a mistake with this pattern in #106: we need to check whether the fragment is already inflated to avoid re-adding it upon configuration changes. See HomeActivity in that PR for how I did this (tl;dr is I added a helper to retrieve the fragment and used that to check presence before reinflation).

File: app/src/main/java/org/oppia/app/player/exploration/ExplorationFragment.kt [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+import android.content.Context
+import android.os.Bundle
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import org.oppia.app.fragment.InjectableFragment
+import javax.inject.Inject
+
+/** Fragment that contains displays single exploration.*/
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/player/exploration/ExplorationFragment.kt]: Ditto here for terminating space.

File: app/src/main/java/org/oppia/app/player/exploration/ExplorationFragmentController.kt [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+import org.oppia.app.fragment.FragmentScope
+import javax.inject.Inject
+
+/** The controller for [ExplorationFragment]. */
+@FragmentScope
+class ExplorationFragmentController @Inject constructor(
+  private val fragment: Fragment
+) {
+  fun handleCreateView(inflater: LayoutInflater, container: ViewGroup?): View? {
+    val binding = ExplorationFragmentBinding.inflate(inflater, container, /* attachToRoot= */ false)
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/player/exploration/ExplorationFragmentController.kt]: Nit: we can probably shorten this to ``return ExplorationFragmentBinding.inflate(...).root``

File: app/src/main/res/layout/exploration_fragment.xml [hunk 7]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="utf-8"?>
+<layout
+  xmlns:android="http://schemas.android.com/apk/res/android">
+  <data>
```
Comment (BenHenning) [inline app/src/main/res/layout/exploration_fragment.xml]: I don't believe we need this data section if it doesn't have anything in it. The outer layout tag is sufficient for generating the data binding subclass.

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @rt4914! Generally LGTM, but I had one concern around the fake data. Other than that, can you also add the initial tests for the exploration activity so that we have a place to add more as the activity grows? At some point, we need to figure out how to divide activity and fragment specific tests, but for now we can just keep UI tests activity-specific.

File: app/src/main/java/org/oppia/app/player/exploration/ExplorationActivity.kt (hunk 2 above)
Comment (rt4914) [reply app/src/main/java/org/oppia/app/player/exploration/ExplorationActivity.kt]: Resolved changed this comment to 
`/** The starting point for exploration. */`

File: app/src/main/java/org/oppia/app/player/exploration/ExplorationFragment.kt (hunk 5 above)
Comment (rt4914) [reply app/src/main/java/org/oppia/app/player/exploration/ExplorationFragment.kt]: Resolved by adding <space> after `.`

File: app/src/main/java/org/oppia/app/player/exploration/ExplorationFragmentController.kt (hunk 6 above)
Comment (rt4914) [reply app/src/main/java/org/oppia/app/player/exploration/ExplorationFragmentController.kt]: Resolved as per suggestion.

File: app/src/main/res/layout/exploration_fragment.xml (hunk 7 above)
Comment (rt4914) [reply app/src/main/res/layout/exploration_fragment.xml]: Thanks. I was not aware of this.
Removed `<data>` tag

File: app/src/main/java/org/oppia/app/player/exploration/ExplorationActivityController.kt (hunk 4 above)
Comment (rt4914) [reply app/src/main/java/org/oppia/app/player/exploration/ExplorationActivityController.kt]: Added `getExplorationFragment()` and add fragment only if it is not present.

Solved as per comment.

File: app/src/main/java/org/oppia/app/player/exploration/ExplorationActivity.kt (hunk 3 above)
Comment (rt4914) [reply app/src/main/java/org/oppia/app/player/exploration/ExplorationActivity.kt]: Changed Fragment and Activity `Controller`s to `Presenter`s

**Review by rt4914 — COMMENTED** [review]
Made changes related to all the files. (Test cases are pending)

File: app/src/main/java/org/oppia/app/FakeDataProvider.kt (hunk 1 above)
Comment (rt4914) [reply app/src/main/java/org/oppia/app/FakeDataProvider.kt]: Actually, I created this file but we are not using it anywhere, so we can remove this right now.

But still there is one issue, we will be using models which are present in `data` module like `GaeSubtitledHtml` `GaeState` etc. in subsequent PRs and as far as I understand the Technical Document, we should not be interacting directly with `data` module from `app` module. So for that case I think `domain` module will need something similar to models defined in `data` module.

Any comments or suggestions on this?

File: app/src/main/java/org/oppia/app/player/state/StateFragment.kt [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+import android.content.Context
+import android.os.Bundle
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import org.oppia.app.fragment.InjectableFragment
+import javax.inject.Inject
+
+/** Fragment that contains displays current state of exploration. */
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/player/state/StateFragment.kt]: Maybe instead:

```kotlin
/** Fragment that represents the current state card of an exploration. */
```

Comment (BenHenning) [reply app/src/main/java/org/oppia/app/player/state/StateFragment.kt]: NB: This is more related to the ``contains displays`` bit, but I think it can be slightly clarified, too.

File: app/src/main/res/layout/exploration_fragment.xml [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+
+  <androidx.constraintlayout.widget.ConstraintLayout
+    android:layout_width="match_parent"
+    android:layout_height="match_parent">
+
+    <TextView
+      android:id="@+id/dummy_text_view"
+      android:layout_width="wrap_content"
+      android:layout_height="wrap_content"
+      android:text="@string/dummy_string"
```
Comment (BenHenning) [inline app/src/main/res/layout/exploration_fragment.xml]: Let's hardcode the string rather than introducing a new string variable. Eventually, we will need to support translating strings, and it's easy to lose track of them. If this is a temporary thing, we can just hardcode the string here & change it in a subsequent PR.

File: app/src/sharedTest/java/org/oppia/app/player/exploration/ExplorationActivityTest.kt [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+import java.util.concurrent.AbstractExecutorService
+import java.util.concurrent.TimeUnit
+import javax.inject.Singleton
+
+/** Tests for [ExplorationActivity]. */
+@RunWith(AndroidJUnit4::class)
+class ExplorationActivityTest {
+  @Before
+  fun setUp() {
+    IdlingRegistry.getInstance().register(MainThreadExecutor.countingResource)
```
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/app/player/exploration/ExplorationActivityTest.kt]: Out of curiosity, is this IdlingResource actually needed for your test? If not, I suggest removing it & the MainThreadExecutor. Prefer not adding extra test infrastructure if not needed.

File: app/src/main/java/org/oppia/app/FakeDataProvider.kt (hunk 1 above)
Comment (BenHenning) [reply app/src/main/java/org/oppia/app/FakeDataProvider.kt]: If this is for testing purposes, we can use the mock GAE services directly in tests to influence the data returned by domain controllers. We can extend those services to actually set up state rather than returning the same fixed data.

If this isn't for tests, app code does not need to rely on GaeSubtitledHtml, etc. since it will receive data via the domain layer defined in the model module (see how topic list data is piped in #106).

Does this help clarify?

**Review by BenHenning — APPROVED** [review]
Thanks @rt4914! This LGTM. Once my last few comments are resolved, feel free to submit.

**Comment by nikitamarysolomanpvt** [comment]
LGTM

File: app/src/main/res/layout/exploration_fragment.xml (hunk 9 above)
Comment (rt4914) [reply app/src/main/res/layout/exploration_fragment.xml]: Makes sense. Will follow this tip from now on.

File: app/src/sharedTest/java/org/oppia/app/player/exploration/ExplorationActivityTest.kt (hunk 10 above)
Comment (rt4914) [reply app/src/sharedTest/java/org/oppia/app/player/exploration/ExplorationActivityTest.kt]: Actually, i am not able to test this because `HomeActivityTest` works first and in that I am getting error

Screenshot 2019-09-17 at 8 26 31 PM

File: app/src/main/java/org/oppia/app/player/state/StateFragment.kt (hunk 8 above)
Comment (rt4914) [reply app/src/main/java/org/oppia/app/player/state/StateFragment.kt]: Resolved and added comment as per @BenHenning suggestion

**Comment by rt4914** [comment]
@BenHenning Check the error related to test cases.

File: app/src/sharedTest/java/org/oppia/app/player/exploration/ExplorationActivityTest.kt (hunk 10 above)
Comment (rt4914) [reply app/src/sharedTest/java/org/oppia/app/player/exploration/ExplorationActivityTest.kt]: Even if I run `ExplorationAcitvityTest`,  `HomeActivityTest` will run and this error occurs.

Comment (BenHenning) [reply app/src/sharedTest/java/org/oppia/app/player/exploration/ExplorationActivityTest.kt]: Ah, this relates to #97. Similarly to how we set up blocking dispatcher, we need to set add these new Logger booleans to the test module for HomeActivityTest so that the're properly set.

Or is this also broken on develop? There's a chance I broke it with #104, but I haven't noticed that yet. It may be the case that this code is now introducing a Logger dependency in HomeActivity, and since HomeActivityTest uses a different root application component we need to make sure all dependencies are satisfied. This will get much easier once #59 is done.

Comment (rt4914) [reply app/src/sharedTest/java/org/oppia/app/player/exploration/ExplorationActivityTest.kt]: I did check that in `develop` right now and it is failing there also.

Comment (BenHenning) [reply app/src/sharedTest/java/org/oppia/app/player/exploration/ExplorationActivityTest.kt]: Ack, thanks! I'll follow up with a fix shortly.

**Comment by BenHenning** [comment]
Ack, will follow up with a fix. Thanks for pointing this out.

File: app/src/sharedTest/java/org/oppia/app/player/exploration/ExplorationActivityTest.kt (hunk 10 above)
Comment (BenHenning) [reply app/src/sharedTest/java/org/oppia/app/player/exploration/ExplorationActivityTest.kt]: #131 is now out for review.

**Comment by BenHenning** [comment]
#131 is now out for review, so sending this back to you to merge once ready.

File: app/src/sharedTest/java/org/oppia/app/player/exploration/ExplorationActivityTest.kt (hunk 10 above)
Comment (rt4914) [reply app/src/sharedTest/java/org/oppia/app/player/exploration/ExplorationActivityTest.kt]: Removed unnecessary code of IdlingResource and MainThreadExecutor

**Comment by rt4914** [comment]
This PR can be merged now and everything has been clarified in respective comments.

**Comment by rt4914** [comment]
> #131 is now out for review, so sending this back to you to merge once ready.

Merging this code now.
